[haassohnpelletstove] Initial contribution

chingon007 opened the pull request:
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->
[haassohnpelletstove]
<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific add-on: Mention the add-on shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->
This binding controls a Pellet stove from Haas & Sohn through the optional WIFI modul. It is able to retreive actual temperature, the actual mode of the stove, the set temperature and turn on and off the stove. The goal was to make this device available in openhab. Some people approached me for openhab support for that stoves. 

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It is a good practice to add a URL to your built JAR in this Pull Request description,
so it is easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

chingon007 commented:
Hello, I did the sign-off as explained. Not sure if it worked correctly.  Any comment on that?

Hilbrand commented:
Thanks for your contribution. The sign-off is checked against each commit and seems to be missing on the first commit. But in general this should be a big issue as it's signed off in the latter commit. Before doing a indept review here 2 general comments. 
1) Use Java code conventions for coding style. Like methods should start with a lower case character, and also cleanup the code. So remove any code commented. See https://www.openhab.org/docs/developer/guidelines.html#java-coding-style. You should run `mvn spotless:apply` on your binding to automatically format the code.
2) Complete the readme.

chingon007 commented:
> Thanks for your contribution. The sign-off is checked against each commit and seems to be missing on the first commit. But in general this should be a big issue as it's signed off in the latter commit. Before doing a indept review here 2 general comments.
> 
> 1. Use Java code conventions for coding style. Like methods should start with a lower case character, and also cleanup the code. So remove any code commented. See https://www.openhab.org/docs/developer/guidelines.html#java-coding-style. You should run `mvn spotless:apply` on your binding to automatically format the code.
> 2. Complete the readme.

That is strange, as I did run several times mvn spotless:apply. Will do it again and check in the changes and update the Readme

Hilbrand commented:
> That is strange, as I did run several times mvn spotless:apply.
It was more of a general comment. Good if you already do this. Sometimes people don't do this. Spotless is mainly for indentation, and won't fix all coding style issues like naming conventions of java methods.

chingon007 commented:
> > That is strange, as I did run several times mvn spotless:apply.
> 
> It was more of a general comment. Good if you already do this. Sometimes people don't do this. Spotless is mainly for indentation, and won't fix all coding style issues like naming conventions of java methods.

Hello, I changed the code to apply to the coding guidelines and added the content to the Readme. Let me know if you still see something missing in order to proceed.

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/README.md:
```diff
@@ -0,0 +1,48 @@
+
+| channel  | type               | description                                              |
+|----------|--------------------|----------------------------------------------------------|
+| prg      | Switch 	 	    | Turn the Stove On/Off		                               |
```
Please use whole words for Channel names. Same for below.

Can you fix the formatting of the table? There are several online markdown table formatters out there.

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/README.md:
```diff
@@ -0,0 +1,48 @@
+|----------|--------------------|----------------------------------------------------------|
+| prg      | Switch 	 	    | Turn the Stove On/Off		                               |
+| isTemp   | Number:Temperature | Receives the actual Temperature of the stove	           |
+| spTemp   | Number:Temperature | Receives and Sets the Target Temperature of the stove	   |
```
```suggestion
| spTemp   | Number:Temperature | Receives and sets the target temperature of the stove	   |
```

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/README.md:
```diff
@@ -0,0 +1,48 @@
+| prg      | Switch 	 	    | Turn the Stove On/Off		                               |
+| isTemp   | Number:Temperature | Receives the actual Temperature of the stove	           |
+| spTemp   | Number:Temperature | Receives and Sets the Target Temperature of the stove	   |
+| mode     | String             | Receives the actual mode the stove is in          	   |
```
Can you mention what states the user can expect?

chingon007 replied:
not really. States depend on oven model and differs from model to model a bit

chingon007 replied:
added an example

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/README.md:
````diff
@@ -0,0 +1,48 @@
+Switch prg    	{ channel="oven:prg" }
+```
+
+## Any custom content here!
+
+_Feel free to add additional sections for whatever you think should also be mentioned about your binding!_
````
Can this be removed?

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/README.md:
````diff
@@ -0,0 +1,48 @@
+
+demo.items:
+
+```java
````
The java keyword shouldn't be necessary.

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/pom.xml:
```diff
@@ -0,0 +1,17 @@
+
+  <artifactId>org.openhab.binding.haassohnpelletoven</artifactId>
+
+  <name>openHAB Add-ons :: Bundles :: haassohnpelletoven Binding</name>
```
Is the capitalization correct?

chingon007 replied:
I guess yes. It was generated automatically. I compared it with other bindings. It's similar

fwolter replied:
Can you use the official notation and capitalization?

chingon007 replied:
Sorry I dont get it. As I said this was automatically generated by the script as described here:

[https://www.openhab.org/docs/developer/](url)

fwolter replied:
Yeah, but that doesn't mean it's correct, isn't it? What about this:

`<name>openHAB Add-ons :: Bundles :: Haas + Sohn Pallet Oven Binding</name>`

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+	Copyright (c) 2010-2020 Contributors to the openHAB project
+
+	See the NOTICE file(s) distributed with this work for additional
+	information.
+
+	This program and the accompanying materials are made available under the
+	terms of the Eclipse Public License 2.0 which is available at
+	http://www.eclipse.org/legal/epl-2.0
+
+	SPDX-License-Identifier: EPL-2.0
+
+-->
```
The `feature.xml` doesn't need a license header. You could remove it.

chingon007 replied:
Will do

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,23 @@
+<features name="org.openhab.binding.haassohnpelletoven-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
+
+	<feature name="openhab-binding-haassohnpelletoven" description="haassohnpelletoven Binding" version="${project.version}">
```
Is the capitalization correct?

chingon007 replied:
again. Generated automatically. Didn't edit here manually

fwolter reviewed (changes requested):
I reviewed the non-Java files until now.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

Please avoid using @SuppressWarnings("null").

Hilbrand commented on bundles/org.openhab.binding.haassohnpelletoven/README.md:
```diff
@@ -0,0 +1,48 @@
+In general two parameters are required. The IP-Address of the WIFI-Modul of the Stove in the local Network and the Access PIN of the Stove.
+The PIN can be found directly at the stove under the Menue/Network/WLAN-PIN
+
+Thing oven  [ hostIP="192.168.0.23", hostPIN=1234]
```
Maybe use 3 backticks, like the demo items.
The thing name should be the full uid, and you can add a label. For example:
`Thing haassohnpelletoven:oven:myOven "Pellet Oven" [ hostIP="192.168.0.23", hostPIN=1234 ]`

chingon007 replied:
will do

chingon007 commented on bundles/org.openhab.binding.haassohnpelletoven/.gitignore:
```diff
@@ -0,0 +1 @@
+/target/
```
Removed

chingon007 commented on bundles/org.openhab.binding.haassohnpelletoven/README.md:
```diff
@@ -0,0 +1,48 @@
+# Haas Sohn Pellet Stove Binding
+
+The binding for Haassohnpelletstove communicates with a Haas and Sohn Pelletstove through the optional
+WIFI-Modul. More information about the WIFI-Modul can be found here: https://www.haassohn.com/de/ihr-plus/WLAN-Funktion
```
Unfortunately not. Seems like the vendor is providing it only in German.

chingon007 reviewed (commented):
ok

chingon007 commented on bundles/org.openhab.binding.haassohnpelletoven/README.md:
```diff
@@ -0,0 +1,48 @@
+
+## Channels
+
+In V0.1 the following channels are yet supported:
```
no, but I will do later enhancements of the code supporting more ovens and more features.

fwolter replied:
The version number of the addon is identical to the version number of the OH distribution. So, v0.1 wouldn't be correct. You could simply write "The following Channels are supported". As mentioned below, roadmaps and plans shouldn't be part of the readme. These can be tracked in GitHub issues.

chingon007 replied:
ok

chingon007 commented on bundles/org.openhab.binding.haassohnpelletoven/README.md:
```diff
@@ -0,0 +1,48 @@
+The PIN can be found directly at the stove under the Menue/Network/WLAN-PIN
+
+Thing oven  [ hostIP="192.168.0.23", hostPIN=1234]
+
+
```
ok

chingon007 commented:
> I reviewed the non-Java files until now.
> 
> There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.
> 
> Please avoid using @SuppressWarnings("null").

removed it

chingon007 commented:
I think I fixed every comment now. Is something in order to merge it finally?

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+        verifyLinkedChannel(CHANNEL_onTime);
+
+        if (!linkedChannels.isEmpty()) {
+            logger.debug("Start automatic refreshing");
```
Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+    @Override
+    public void dispose() {
+
+        logger.debug("Disposing Haas and Sohn Pellet stove handler.");
```
See above. Debugger.

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+            Runnable runnable = new Runnable() {
+                @Override
+                public void run() {
+                    try {
+
+                        updateOvenData(null);
+
+                        for (Channel channel : getThing().getChannels()) {
+                            updateChannel(channel.getUID().getId());
+                        }
+                    } catch (Exception e) {
+                        logger.debug("Exception occurred during execution: {}", e.getMessage(), e);
+                    }
```
You could make this a private method.

fwolter replied:
I meant the code, which is currently in the anonymous inner class method `run()`. This can be moved into a private method. Then, the Runnable could be removed.

chingon007 replied:
my intention is to run this in an extra Thread. According to [https://docs.oracle.com/javase/7/docs/api/java/lang/Runnable.html](url) this is the correct way.

fwolter replied:
You're on the right way! The code can be simplified if you use lambda expressions. This can be achieved by moving this code in a separate method:
```
private void myMethod() {
                    try {

                        updateOvenData(null);

                        for (Channel channel : getThing().getChannels()) {
                            updateChannel(channel.getUID().getId());
                        }
                    } catch (Exception e) {
                        logger.debug("Exception occurred during execution: {}", e.getMessage(), e);
                    }
}
```
Then, you can do
`refreshJob = scheduler.scheduleWithFixedDelay(this::myMethod, 0, period, TimeUnit.SECONDS);`

chingon007 replied:
I understand. I do know Lambda expressions from C#. In general I am not sure if all of this little points are worth the effort. As I found the exact same code snippet in a other already available binding.
What I mean, what is now really holding the source code back to be part of main? In general the binding works very well for my purpose in my openhab already since weeks. There are a lot of users who would love to have it and approached me when will it be part of openhab. I cannot answer that. I am loosing a bit the desire about this whole topic if it is so effort intense to contribute a binding, a lot can use, to openhab. 
Don't get me wrong, some code review findings have been helpful and are indeed correct, but it seems like an infinite journey here. I am considering to close the pull without merging it.........
Let me know what you think.

fwolter replied:
Great that you got the point with the lambda expression! The openHAB addon repo has a quite large code base. So, key is to keep the code maintainable and see that new code is in a good shape before it gets merged. If there are improvements that make the code more concise and can easily applied, it should be done. That is gonna be crucial if somebody else needs to edit your code. That is important for changes that affect all add-ons, like API changes in the core.

A review takes normally 2-3 loops, before another maintainer starts with the second review. The latter has less cycles in general.

chingon007 replied:
Changed it.

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+
+            @SuppressWarnings("null")
+            int period = config.refreshRate;
+            refreshJob = scheduler.scheduleWithFixedDelay(runnable, 0, period, TimeUnit.SECONDS);
```
Then, you can call it here:
```suggestion
            refreshJob = scheduler.scheduleWithFixedDelay(this::myMethod, 0, period, TimeUnit.SECONDS);
```

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+                        for (Channel channel : getThing().getChannels()) {
+                            updateChannel(channel.getUID().getId());
+                        }
+                    } catch (Exception e) {
```
What type of exception do you expect here? Can you specify the concrete type?

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+                }
+            };
+
+            @SuppressWarnings("null")
```
See above. Could be removed.

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+    public void channelLinked(ChannelUID channelUID) {
+
+        if (!automaticRefreshing) {
+            logger.debug("Start automatic refreshing");
```
See above. Debugger.

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+        if (linkedChannels.isEmpty()) {
+            automaticRefreshing = false;
+            stopScheduler();
+            logger.debug("Stop automatic refreshing");
```
See above. Debugger.

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+    @Override
+    public void channelUnlinked(ChannelUID channelUID) {
+
+        if (linkedChannels.containsKey(channelUID.getId())) {
+
+            linkedChannels.remove(channelUID.getId());
+        }
```
No harm in removing a key that doesn't exist.
```suggestion
            linkedChannels.remove(channelUID.getId());
```

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+                        update(state, channelId);
+                        break;
+                    case CHANNEL_prg:
+                        boolean prg = data.getPrg();
+                        if (prg) {
+                            state = OnOffType.ON;
+                        } else {
+                            state = OnOffType.OFF;
+                        }
+                        update(state, channelId);
```
Syntactical sugar. Same for below.
```suggestion
                        update(OnOffType.from(data.getPrg()), channelId);
```

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+                        update(state, channelId);
+                        break;
+                    case CHANNEL_spTemp:
+                        state = new StringType(data.getspTemp());
```
You need to use `QuantityType` here. Same for below.
```suggestion
                        state = new QuantityType<Temperature>(Double.valueOf(data.getspTemp()), SIUnits.CELSIUS);
```

chingon007 replied:
ok, but not as you suggested. throws error.

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/validation/IpAddressValidator.java:
```diff
@@ -0,0 +1,43 @@
+@NonNullByDefault
+public class IpAddressValidator {
+
+    private static final String zeroTo255 = "([01]?[0-9]{1,2}|2[0-4][0-9]|25[0-5])";
+
+    private static final String IP_REGEXP = zeroTo255 + "\\." + zeroTo255 + "\\." + zeroTo255 + "\\." + zeroTo255;
+
+    private static final Pattern IP_PATTERN = Pattern.compile(IP_REGEXP);
```
Actually the framework takes care that a valid IP address or hostname is specified. Is there a reason why you make an additional validation here?

chingon007 replied:
no, I didn't know that. I would like to keep it anyway to assure a right error messaging and it is working though.

fwolter replied:
The UI will produce a correct error message. So, your code would be redundant. For the sake of better code maintainability, it's better to remove it.

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/validation/PinValidator.java:
```diff
@@ -0,0 +1,39 @@
+@NonNullByDefault
+public class PinValidator {
+
+    Pattern pattern = Pattern.compile("(\\d{4})");
```
You could specify the pattern in the XML file. I made a suggestion below.

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+
+	<name>Haas and Sohn Pelletoven Binding</name>
+	<description>The binding for Haassohnpelletstove communicates with a Haas and Sohn Pelletstove through the optional
+		WLAN-Modul. More information can be found here: https://www.haassohn.com/de/ihr-plus/WLAN-Funktion</description>
```
I think the link won't be converted to a link. So, you could remove it here as it is already contained in the readme.
```suggestion
		WIFI module. More information can be found here: https://www.haassohn.com/de/ihr-plus/WLAN-Funktion</description>
```

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>Haas and Sohn Pelletoven Binding</name>
+	<description>The binding for Haassohnpelletstove communicates with a Haas and Sohn Pelletstove through the optional
```
```suggestion
	<description>This binding communicates with Haas and Sohn Pelletstoves through the optional
```

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/i18n/haassohnpelletoven_xx_XX.properties:
```diff
@@ -0,0 +1,29 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
+# FIXME: please do not add the file to the repo if you add or change no content
```
You need to replace the xx in the filename by the locale you want to cover.

As the english locale is already covered by the XML file, this file could be removed.

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,128 @@
+
+	<!-- Haas Sohn Thing Type -->
+	<thing-type id="oven">
+		<label>Haas and Sohn Pellet stove Binding Thing</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,128 @@
+	<!-- Haas Sohn Thing Type -->
+	<thing-type id="oven">
+		<label>Haas and Sohn Pellet stove Binding Thing</label>
+		<description>The binding for haassohnpelletstove communicates with a Haas and Sohn Pelletoven through the optional
+			WLAN-Modul. More information can be found here: https://www.haassohn.com/de/ihr-plus/WLAN-Funktion</description>
```
This would apply to the whole binding. Can you express this a bit different to describe the Thing?

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,128 @@
+			<channel id="channelIsTemp" typeId="isTemp"/>
+			<channel id="channelMode" typeId="mode"/>
+			<channel id="channelSpTemp" typeId="spTemp"/>
+			<channel id="channelPrg" typeId="prg"/>
```
You could use the predefined Channel `system.power`. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#system-state-channel-types

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,128 @@
+
+		<config-description>
+			<parameter name="hostIP" type="text" required="true">
+				<label>IP Address</label>
```
You could specify `<context>network-address</context>` to get a free validation.

chingon007 replied:
keeping mine

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,128 @@
+				<label>IP Address</label>
+				<description>Please add the IP Address of the WIFI Module of the Haas and Sohn oven here</description>
+			</parameter>
+			<parameter name="hostPIN" type="text" required="true">
```
```suggestion
			<parameter name="hostPIN" type="text" required="true" pattern="[0-9]{4}">
```

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,128 @@
+				<label>PIN</label>
+				<description>Please add the PIN of your oven here. You can find it in the Menu directly in your oven.</description>
+			</parameter>
+			<parameter name="refreshRate" type="integer">
```
```suggestion
			<parameter name="refreshRate" type="integer" unit="s" min="0" max="1000">
```

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,128 @@
+				<label>Refresh Rate</label>
+				<description>
+					How often the Pellet Stove should schedule a refresh after a channel is linked to an item. Temperature
+					data will be
+					refreshed according this set time in seconds. Valid input is 0 - 999.
```
Can you fix the line break?

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,128 @@
+        <item-type>Switch</item-type>
+        <label>On/Off for the Eco Mode of the stove</label>
+        <description>To turn the stove on/off as switch</description>
+        <event>
+            <options>
+                <option value="ON">ON</option>
+                <option value="OFF">OFF</option>
+            </options>
+        </event>
```
A switch doesn't need event options.

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,128 @@
+    
+    <channel-type id="ignitions">
+        <item-type>String</item-type>
+        <label>Amount of ignition's of the stove</label>
```
See above label length. Please check all.
```suggestion
        <label>Amount of ignitions of the stove</label>
```

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,128 @@
+    
+    <channel-type id="maintenanceIn">
+        <item-type>String</item-type>
+        <label>States the next maintenance in kg</label>
```
Words in labels should be capitalized (except prepositions and so on). See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.lgwebos/README.md:
```diff
@@ -16,7 +16,7 @@ Under network settings allow "LG CONNECT APPS" to connect.
 Note: Under general settings allow mobile applications to turn on the TV, if this option is available.
 On newer models this setting may also be called "Mobile TV On > Turn On Via WiFi".
-In combination with the wake on LAN binding this will allow you to start the TV via openHAB. Please see demo.items and demo.rules example below.
+If channel `power` receives `ON`, the binding will attempt to power on the TV by broadcasting a Wake on Lan packet.
```
Can you revert this change?

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,128 @@
+
+	</thing-type>
+	<channel-type id="isTemp">
+		<item-type>String</item-type>
```
Please check all.
```suggestion
		<item-type>Number:Temperature</item-type>
```

fwolter reviewed (changes requested):
There are some formatting issues. You can fix them with `mvn spotless:apply`.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

At some places you used haasohnpelletstove and at others haasohnpelletoven.

All of you packages should be located in the `internal` package.

You need to sign-off your commits. See https://www.openhab.org/docs/developer/contributing.html#sign-your-work The necessary commands are listed when you click on "Details" at the DCO check below.

Better don't use `git merge` in pull request as this can garble your Git history when fixing the DCO check.

Just to manage your expectations, as this is a new binding, another maintainer needs to review it.

chingon007 reviewed (commented):
Ok, I checked and applied most of it. About to upload the modified files. Had problem with the signature in Git. Hope it is fixed now.

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/communication/HaassohnpelletstoveJSONCommunication.java:
```diff
@@ -0,0 +1,249 @@
+                    + ". Cannot find service on give IP. Please verify the IP-Address!";
+            errorDetail = e.getMessage();
+            resultOk = false;
+        } catch (Exception e) {
```
If you miss an unchecked exception the framework takes care of it to be logged. So, this could be removed.

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/communication/HaassohnpelletstoveJSONCommunication.java:
```diff
@@ -0,0 +1,249 @@
+
+        if (postData != null) {
+
+            InputStream targetStream = new ByteArrayInputStream(postData.getBytes());
```
If the encoding is not specified, the default encoding of the JVM will be used. As this is a user setting, your code can behave differently on different systems if it is not specified.

chingon007 replied:
ok. added encoding.

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/data/HaassohnpelletstoveJsonData.java:
```diff
@@ -0,0 +1,143 @@
+ *
+ * @author Christian Feininger - Initial contribution
+ */
+@NonNullByDefault
```
You can simply remove the `@NonNullByDefault` annotation.

chingon007 replied:
if I do this I will get a Warning during build:

`[WARNING] org.openhab.binding.haassohnpelletstove.internal.HaasSohnpelletstoveJsonData.java:[25]
Classes/Interfaces should be annotated with @NonNullByDefault`

and also a Prio 2 checkstyle fault in the code-analysis report.

fwolter replied:
You can add "DTO" in the name of this class to avoid a checkstyle warning.

chingon007 replied:
Now I understood what you meant. Will do so. Didn't know this will be considered

fwolter commented on bundles/org.openhab.binding.haassohnpelletoven/src/main/java/org/openhab/binding/haassohnpelletoven/internal/HaassohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,390 @@
+                QuantityType<?> value = null;
+                try {
+                    value = (QuantityType<?>) command;
+                    double a = value.doubleValue();
```
Users can configure their temperature in Celsius and Fahrenheit in the UI. If a user configured Fahrenheit, the value returned by `.doubleValue()` will be in Fahrenheit. You need to convert it to Celsius. Otherwise the oven will be fed by the Fahrenheit value.

chingon007 replied:
ok

chingon007 commented:
Hello,

after refactoring and naming everything to pelletstove instead of oven. I get an error on command line when I build it again:

``
[ERROR] [ERROR] Could not find the selected project in the reactor: :org.openhab.binding.haassohnpelletstove @
[ERROR] Could not find the selected project in the reactor: :org.openhab.binding.haassohnpelletstove -> [Help 1]
org.apache.maven.MavenExecutionException: Could not find the selected project in the reactor: :org.openhab.binding.haassohnpelletstove
    at org.apache.maven.graph.DefaultGraphBuilder.trimSelectedProjects (DefaultGraphBuilder.java:172)
    at org.apache.maven.graph.DefaultGraphBuilder.reactorDependencyGraph (DefaultGraphBuilder.java:124)
    at org.apache.maven.graph.DefaultGraphBuilder.build (DefaultGraphBuilder.java:84)
    at org.apache.maven.DefaultMaven.buildGraph (DefaultMaven.java:507)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:219)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[ERROR]
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MavenExecutionException
``

I did the appropriate change in the bundles pom.xml. What am I missing @fwolter ?

fwolter commented:
Just to avoid confusion, I didn't meant to remove your package structure at all, but to move all your packages in the `internal` one. But a flat hierarchy like now is ok as the binding hasn't that much classes.

chingon007 commented:
> Just to avoid confusion, I didn't meant to remove your package structure at all, but to move all your packages in the `internal` one. But a flat hierarchy like now is ok as the binding hasn't that much classes.

its ok. but I have no problems with building on command line. Any idea what am I missing?

fwolter commented:
There are two pom.xml. You need to add your binding to `bom\openhab-addons\pom.xml`, too.

chingon007 commented:
> There are two pom.xml. You need to add your binding to `bom\openhab-addons\pom.xml`, too.

Doesn't fix it either

fwolter commented:
The binding compiles on my computer. You could push your current changes. I will take a look into it, then.

chingon007 commented:
> The binding compiles on my computer. You could push your current changes. I will take a look into it, then.

I did now. I wanted to have a look on the ``target/code-analysis/report.html.`` but without complining I can continue here

fwolter commented:
I can compile the binding. What commands do you execute in which directory?

Please note that the sign-off message must contain your real name.

chingon007 commented:
> I can compile the binding. What commands do you execute in which directory?
> 
> Please note that the sign-off message must contain your real name.

In ``openhab-addons`` folder I exectue ``mvn clean install -pl :org.openhab.binding.haassohnpelletstove``

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/pom.xml:
```diff
@@ -0,0 +1,17 @@
+    <version>3.1.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>org.openhab.binding.haassohnpellestove</artifactId>
```
That's the reason why you can't compile it:
```suggestion
  <artifactId>org.openhab.binding.haassohnpelletstove</artifactId>
```

chingon007 replied:
wow, that made me nuts. Thanks @fwolter you saved my day. Will do the rest of the suggested changes. 
It does compile again.

chingon007 replied:
There are even more typos with it in other files I had to fix. thanks for help. It runs again.

chingon007 commented:
> There are some formatting issues. You can fix them with `mvn spotless:apply`.
> 
> There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.
> 
> At some places you used haasohnpelletstove and at others haasohnpelletoven.
> 
> All of you packages should be located in the `internal` package.
> 
> You need to sign-off your commits. See https://www.openhab.org/docs/developer/contributing.html#sign-your-work The necessary commands are listed when you click on "Details" at the DCO check below.
> 
> Better don't use `git merge` in pull request as this can garble your Git history when fixing the DCO check.
> 
> Just to manage your expectations, as this is a new binding, another maintainer needs to review it.

ok

chingon007 commented:
> There are some formatting issues. You can fix them with `mvn spotless:apply`.
> 
> There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.
> 
> At some places you used haasohnpelletstove and at others haasohnpelletoven.
> 
> All of you packages should be located in the `internal` package.
> 
> You need to sign-off your commits. See https://www.openhab.org/docs/developer/contributing.html#sign-your-work The necessary commands are listed when you click on "Details" at the DCO check below.
> 
> Better don't use `git merge` in pull request as this can garble your Git history when fixing the DCO check.
> 
> Just to manage your expectations, as this is a new binding, another maintainer needs to review it.

ok

chingon007 commented:
Hello,
I am done with fixing all review findings @fwolter . How do we proceed now in order to merge it?

Hilbrand commented:
There needs to be a second review and when both reviews approve the binding will be merged. I've some pending reviews (not visible yet) I plan to give soon. In the mean time there are some additional commits in this pull request that should not be in there. Can you remove them? Or if you don't know how to do it I can remove them.

chingon007 commented:
> There needs to be a second review and when both reviews approve the binding will be merged. I've some pending reviews (not visible yet) I plan to give soon. In the mean time there are some additional commits in this pull request that should not be in there. Can you remove them? Or if you don't know how to do it I can remove them.

Thanks for the info. Go ahead and delete the commits  which should not be in there. I don't know which one you are in detail refering to.

Hilbrand commented:
I've fixed your branch. I've done a push force so this mean you can't just do a pull because that would break your local branch. To update your local branch you need to do the following:
First get the updated remote branch:
```
git fetch origin
```
Then you need to update your local main branch with the following command:
If you have any local changes not yet committed you can use reset with --soft
```
git reset --soft
```

Hilbrand commented:
I realized your local git is probably a standard openHAB eclipse installation and then the instructions to update need to be different. I've fixed your branch. I've done a push force so this mean you can't just do a pull because that would break your local branch. To update your local branch you need to do the following.

```
git remote rename origin upstream
git remote add origin https://github.com/chingon007/openhab-addons.git
```
If you have any local changes not yet committed you can use reset with --soft
```
git reset --soft origin/main
```

chingon007 commented:
> I realized your local git is probably a standard openHAB eclipse installation and then the instructions to update need to be different. I've fixed your branch. I've done a push force so this mean you can't just do a pull because that would break your local branch. To update your local branch you need to do the following.
> 
> ```
> git remote rename origin upstream
> git remote add origin https://github.com/chingon007/openhab-addons.git
> ```
> 
> If you have any local changes not yet committed you can use reset with --soft
> 
> ```
> git reset --soft origin/main
> ```
Thanks for the Info. I got my local branch working again (at least I hope so) and pushed a minor change in the Readme. Let me know if this is ok now the way it is.

fwolter commented:
I don't know which commit to the readme you are refering to, but your branch looks good.

chingon007 commented:
Ok, so lets merge or what is holding it back?

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/README.md:
```diff
@@ -0,0 +1,67 @@
+The following channels are yet supported:
+
+
+| Channel | Type  | Access| Description|
+|---------|-------|-------|------------|
+| power| Switch | read/write|Turn the stove on/off|
+|channelIsTemp|Number:Temperature|read|Receives the actual temperature of the stove|
+|channelSpTemp|Number:Temperature|read/write|Receives and sets the target temperature of the stove|
```
Can you fix the formatting of the table? There are several online markdown table formatters on the net.

chingon007 replied:
Fixed it before. Looks perfect already in the repo. Have a look here [https://github.com/chingon007/openhab-addons/tree/main/bundles/org.openhab.binding.haassohnpelletstove](url)

fwolter replied:
The link you posted is broken. I see it still unformatted in your branch: https://raw.githubusercontent.com/chingon007/openhab-addons/main/bundles/org.openhab.binding.haassohnpelletstove/README.md

chingon007 replied:
This is indeed correct. If you open it in a browser, it looks perfect (see screenshot). This is how I got it out of a online formatter. 
![2021-02-21 10_05_24-openhab-addons_README md at main · chingon007_openhab-addons](https://user-images.githubusercontent.com/76529461/108620448-96a96900-742c-11eb-99ad-44651e1c28f6.png)

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/README.md:
````diff
@@ -0,0 +1,67 @@
+Switch power { channel="oven:power" }
+```
+
+## Google Assistant configuation
+See also: https://www.openhab.org/docs/ecosystem/google-assistant/
````
There must be an empty line between heading and text in markdown.

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/src/main/java/org/openhab/binding/haassohnpelletstove/internal/HaasSohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,327 @@
+            }
+        } else if (channelUID.getId().equals(CHANNELSPTEMP)) {
+            if (command instanceof QuantityType<?>) {
+                @SuppressWarnings("unchecked")
+                QuantityType<Temperature> value = (QuantityType<Temperature>) command;
```
You can avoid the supression by using the wildcard.
```suggestion
                QuantityType<?> value = (QuantityType<?>) command;
```

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/src/main/java/org/openhab/binding/haassohnpelletstove/internal/HaasSohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,327 @@
+                    updateLinkedChannels();
+                }
+            } else {
+                logger.info("Setting thing '{}' to OFFLINE: {}", getThing().getUID(), errors);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, message.getStatusDesription());
```
The log message could be removed, as the status update is already logged by the framework.

The state change originated by updateStatus() is logged to events.log. Including the status detail message.

Please check all.

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/src/main/java/org/openhab/binding/haassohnpelletstove/internal/HaasSohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,327 @@
+        verifyLinkedChannel(CHANNELCONSUMPTION);
+        verifyLinkedChannel(CHANNELONTIME);
+        if (!linkedChannels.isEmpty()) {
+            logger.info("Start automatic refreshing");
```
Logging to info should be used sparsely e.g. a newly started component or a user file that has been loaded. This could be trace.

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/src/main/java/org/openhab/binding/haassohnpelletstove/internal/HaasSohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,327 @@
+
+    @Override
+    public void dispose() {
+        logger.info("Disposing Haas and Sohn Pellet stove handler.");
```
Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/src/main/java/org/openhab/binding/haassohnpelletstove/internal/HaasSohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,327 @@
+            for (Channel channel : getThing().getChannels()) {
+                updateChannel(channel.getUID().getId());
+            }
+        } catch (Exception e) {
```
What type of exception do you expect here? Can you specify the concrete type?

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/src/main/java/org/openhab/binding/haassohnpelletstove/internal/MD5Utils.java:
```diff
@@ -0,0 +1,66 @@
+        return result;
+    }
+
+    private static String bytesToHex(byte[] bytes) {
+        StringBuilder sb = new StringBuilder();
+        for (byte b : bytes) {
+            sb.append(String.format("%02x", b));
+        }
+        return sb.toString();
+    }
```
The framework has a method for converting bytes to hex String. To make the code better maintainable, please take a look at https://openhab.org/javadoc/latest/org/openhab/core/util/hexutils

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,11 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>Haas and Sohn Pelletstove Binding</name>
+	<description>This binding communicates with Haas and Sohn Pelletstoves through the optional
+		WLAN-Modul. It allows to
+		power the stove on and off and receives different operation information.</description>
```
Can you fix the line break?
```suggestion
	<description>This binding communicates with Haas and Sohn Pelletstoves through the optional
		WIFI module. It allows to
		power the stove on and off and receives different operation information.</description>
```

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/src/main/resources/OH-INF/i18n/haassohnpelletstove_xx_XX.properties:
```diff
@@ -0,0 +1,29 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
```
You need this file if you want to translate the binding to another language than English. As you have only English, this file could be removed.

chingon007 replied:
ok

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,117 @@
+
+	<!-- Haas Sohn Thing Type -->
+	<thing-type id="oven">
+		<label>Haas and Sohn Pelletstove Binding</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions
```suggestion
		<label>Haas+Sohn Oven</label>
```

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,117 @@
+	<!-- Haas Sohn Thing Type -->
+	<thing-type id="oven">
+		<label>Haas and Sohn Pelletstove Binding</label>
+		<description>The binding for Haas and Sohn Pelletstove communicates with a Haas and Sohn Pelletstove through the
+			optional
+			WLAN-Modul. More information can be found here: https://www.haassohn.com/de/ihr-plus/WLAN-Funktion. It allows
+			to power on/off the stove as well as receiving different operation information about the stove.</description>
```
This should be the description for the Thing. You are refering to the binding.

chingon007 replied:
ok

fwolter reviewed (changes requested):
There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

chingon007 commented:
tried to fix everything

chingon007 commented:
Any update here in order to finally merge? Anything pending?

fwolter commented on bundles/org.openhab.binding.haassohnpelletstove/src/main/java/org/openhab/binding/haassohnpelletstove/internal/HaasSohnpelletstoveHandler.java:
```diff
@@ -0,0 +1,317 @@
+        if (linkedChannels.isEmpty()) {
+            automaticRefreshing = false;
+            stopScheduler();
+            logger.info("Stop automatic refreshing");
```
Logging to info should be used sparsely e.g. a newly started component or a user file that has been loaded. This could be debug or removed at all.

chingon007 replied:
ok, changed it.

fwolter commented on bundles/org.openhab.binding.openthermgateway/README.md:
```diff
@@ -32,41 +32,47 @@ The configuration settings for the thing are Hostname/IP address and Port, which
 
 The OpenTherm Gateway binding supports the following channels:
 
-| Channel Type ID          | Item Type            | Description                                              | Read Only |
```
Please remove the changes of other bindings from your PR.

chingon007 replied:
I am not sure how this ended up in my pull, tried to remove it.

fwolter commented:
I wanted to remove the openthermgateway commit from your branch, but I made a mistake with the git command arguments. Now, I don't have permission to push to your main branch anymore. If you re-add the permission in your GitHub branch settings, I will fix it.

chingon007 commented:
You should have the wrights again. However I did not revoke your rights.
Also my Pull request is now closed and I cant reopen it. Did you close it by mistake @fwolter ?

fwolter commented:
The PR was closed automatically. Same for the permissions. I don't have write access to your repo. You probably need to invite me as a collaborator via your repo settings -> manage access -> invite.

chingon007 commented:
Ok, you should have access now. Let me know what I can do in order to proceed.

fwolter commented:
GitHub doesn't detect any differences between your main branch and the upstream branch. That might be caused due to the removal of the openthermgateway commit, which was the very first commit in your PR. Can you please open a new PR from your main branch?

chingon007 commented:
Sorry, you need me help out a bit here in order to do the right thing. Do I need to create a new Pull Request from openhab:main like I did on this on and integrate there my code again?

fwolter commented:
You simply need to create a new PR from the main branch of your repo. Your new code will show up automatically in this new PR.

chingon007 commented:
I did. I am not sure if it worked as this pull is still closed?

fwolter commented:
You need to create a new PR, like you did for this. This PR won't open again. Your code will show up in the newly created PR.
